refactor: Simplify CTxMempool/BlockAssembler fields, remove some external mapTx access

sedited opened the pull request:
Motivation
* It seems preferable to use stdlib data structures instead of boost if they can achieve close to the same thing.
* Code external to mempool should ideally use its public helper methods instead of accessing `mapTx` or its iterators directly.
* Reduce the number of complex boost multi index type interactions
* Also see #28335 for further context/motivation. This PR together with #28385 simplifies that one.

Overview of things done in this PR:
* Make `vTxHashes` a vector of transaction references instead of a pair of transaction hash and iterator. The trade off here is that the data is retrieved on the fly with `GetEntry` instead of being cached in `vTxHashes`.
* Introduce `GetEntry` helper method to replace the more involved `GetIter` where applicable
* Replace `mapTx` access with `CTxMemPool` helper methods
* Simplify `checkChainLimits` call in `node/interfaces.cpp`
* Make `CTxMemPoolEntry`s `lockPoints`mutable such that they can be changed with a const iterator directly instead of going through `mapTx`
* Make `BlockAssembler`'s `inBlock` and `failedTx` sets of transaction hashes.

glozow committed:
[refactor] get wtxid from entry instead of vTxHashes

glozow committed:
[refactor] rewrite vTxHashes as a vector of CTransactionRef

vTxHashes exposes a complex mapTx iterator type that its external users
don't need. Directly populate it with CTransactionRef instead.

glozow committed:
[refactor] rewrite BlockAssembler inBlock and failedTx as sets of txids

glozow committed:
[refactor] remove access to mapTx from rpc/mempool.cpp

glozow committed:
[refactor] remove access to mapTx from policy/rbf.cpp

glozow committed:
[refactor] remove access to mapTx in blockencodings_tests.cpp

glozow committed:
[refactor] remove access to mapTx.find in mempool_tests.cpp

glozow committed:
[refactor] use exists() instead of mapTx.find()

glozow committed:
[refactor] use CheckPackageLimits for checkChainLimits

The behavior is the same as CalculateMemPoolAncestors. The only
difference is the string returned, and the string is discarded anyway
since checkChainLimits only cares about pass/fail.

glozow committed:
[txmempool] make CTxMemPoolEntry::lockPoints mutable

Allows calling UpdateLockPoints() with a (const) txiter. Note that this
was already possible for caller using mapTx.modify(txiter). The point
here is to not be accessing mapTx when doing so.

glozow committed:
[refactor] remove access to mapTx in validation.cpp

sedited committed:
scripted-diff: rename vTxHashes to txns_randomized

-BEGIN VERIFY SCRIPT-
git grep -l "vTxHashesIdx" src | xargs sed -i "s/vTxHashesIdx/idx_randomized/g"
git grep -l "vTxHashes" src | xargs sed -i "s/vTxHashes/txns_randomized/g"
-END VERIFY SCRIPT-

glozow commented:
Concept ACK fwiw
As additional motivation, I imagine this would also make it easier to change the internal implementation of `CTxMemPool` without touching all these files.

ismaelsadeeq commented on src/node/interfaces.cpp:
```diff
@@ -698,11 +697,9 @@ class ChainImpl : public Chain
         LOCK(m_node.mempool->cs);
-        return m_node.mempool->CalculateMemPoolAncestors(entry, limits).has_value();
+        std::string err_string;
+        return m_node.mempool->CheckPackageLimits({tx}, m_node.mempool->m_limits, err_string);
```
CI failure is due to merge conflict can be fixed with
```suggestion
        return m_node.mempool->CheckPackageLimits({tx}, GetVirtualTransactionSize(*tx), err_string);
```

ismaelsadeeq commented on src/test/blockencodings_tests.cpp:
```diff
@@ -51,8 +51,8 @@ static CBlock BuildBlockTestCase() {
 // Number of shared use_counts we expect for a tx we haven't touched
-// (block + mempool + our copy from the GetSharedTx call)
-constexpr long SHARED_TX_OFFSET{3};
+// (block + mempool entry + mempool txns_randomized + our copy from the GetSharedTx call)
```
In b0dda173e1784f177152349ff7002345f054f888 [refactor] rewrite vTxHashes as a vector of CTransactionRef 
`txns_randomized` is introduced in next commit d7eec401ba7c7b4c8fc58e39e358ba4518da97d7

Does this belongs to this commit?

ismaelsadeeq reviewed (commented):
Concept ACK on the changes on in this PR.

This refactor all seems relevant.
The commits are really atomic and easier to review. 💯

TheCharlatan commented:
Rebased a5bdd7eacea7be11e064aa7a9a4f12624dfe53fd -> f0577ea4c428421a85e98bfeb2fbd5e617a4622a ([simplifyMemPoolInteractions_0](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_0) -> [simplifyMemPoolInteractions_1](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_1),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_0..simplifyMemPoolInteractions_1))

TheCharlatan commented:
Rebased f0577ea4c428421a85e98bfeb2fbd5e617a4622a -> 5fb9ddf0ab423dae5c0056eda10b02817927fcc4 ([simplifyMemPoolInteractions_1](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_1) -> [simplifyMemPoolInteractions_2](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_2),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_1..simplifyMemPoolInteractions_2))

ismaelsadeeq commented on src/node/interfaces.cpp:
```diff
@@ -700,11 +699,9 @@ class ChainImpl : public Chain
-        LockPoints lp;
-        CTxMemPoolEntry entry(tx, 0, 0, 0, 0, false, 0, lp);
-        const CTxMemPool::Limits& limits{m_node.mempool->m_limits};
         LOCK(m_node.mempool->cs);
```
The change in https://github.com/bitcoin/bitcoin/commit/e7e462eadc66e484ccb2f51967a0ad5f0d3bd41a has been reverted after the rebase.
```suggestion
        CTxMemPoolEntry entry(tx, 0, 0, 0, 0, false, 0, lp);
        LOCK(m_node.mempool->cs);
        std::string err_string;
        return m_node.mempool->CheckPackageLimits({tx}, entry.GetTxSize(), err_string);
```

TheCharlatan commented:
Updated 5fb9ddf0ab423dae5c0056eda10b02817927fcc4 -> e579bb98ba8977af284ba6914ffb2b1da7f34cdd ([simplifyMemPoolInteractions_2](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_2) -> [simplifyMemPoolInteractions_3](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_3),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_2..simplifyMemPoolInteractions_3))

* Addressed @ismaelsadeeq's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1335223808), fixed change ordering in commit (no net change to the code).
* Fixed reverted change by cherry-picking commit e7e462eadc66e484ccb2f51967a0ad5f0d3bd41a [refactor] use CheckPackageLimits for checkChainLimits.

ismaelsadeeq commented on src/policy/rbf.cpp:
```diff
@@ -35,7 +36,7 @@ RBFTransactionState IsRBFOptIn(const CTransaction& tx, const CTxMemPool& pool)
     // If all the inputs have nSequence >= maxint-1, it still might be
     // signaled for RBF if any unconfirmed parents have signaled.
-    const CTxMemPoolEntry entry{*pool.mapTx.find(tx.GetHash())};
+    const CTxMemPoolEntry entry{**Assert(pool.GetIter(tx.GetHash()))};
```
Are you assuming that the callers of `IsRBFOptIn` must pass a mempool transaction with the `Assert`?

TheCharlatan replied:
Yes, as I understand it that is what is checked in the code block above, so this should always hold.

ismaelsadeeq reviewed (commented):
Light Code review ACK  e579bb98ba8977af284ba6914ffb2b1da7f34cdd
Non expert look at the changes here,  there are no any external usage of `CTxMemPool` `mapTx.find`.
 The changes here looks good to me

Just a question and nits below.
I see there are still some external access of `mapTx` in the codebase like ~~`mapTx.size`~~, `mapTx.get` and `mapTx.project`.

sedited committed:
[refactor] remove access to mapTx in validation_block_tests

Use the helper function instead of reaching into the mapTx member
object.

TheCharlatan commented on src/rpc/mempool.cpp:
```diff
@@ -315,8 +315,8 @@ static void entryToJSON(const CTxMemPool& pool, UniValue& info, const CTxMemPool
     UniValue spent(UniValue::VARR);
-    const CTxMemPool::txiter& it = pool.mapTx.find(tx.GetHash());
-    const CTxMemPoolEntry::Children& children = it->GetMemPoolChildrenConst();
+    const auto it = pool.GetIter(tx.GetHash());
+    const CTxMemPoolEntry::Children& children = (*CHECK_NONFATAL(it))->GetMemPoolChildrenConst();
```
Huh, yeah. That is more straight forward.

TheCharlatan commented:
Thank you for the review @ismaelsadeeq!

Updated e579bb98ba8977af284ba6914ffb2b1da7f34cdd -> 105a0f4db4ffdc25d3ad30300c949d46d5d8e647 ([simplifyMemPoolInteractions_3](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_3) -> [simplifyMemPoolInteractions_4](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_4),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_3..simplifyMemPoolInteractions_4))

* Added commit 105a0f4db4ffdc25d3ad30300c949d46d5d8e647 for removing access to `mapTx` in `validation_block_tests`.
* Addressed @ismaelsadeeq's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1378721448), simplifying code by just using mempool entry directly.
* Addressed @ismaelsadeeq's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1378747401), making commit message more precise.

ismaelsadeeq commented:
Re ACK 105a0f4db4ffdc25d3ad30300c949d46d5d8e647

ismaelsadeeq commented on src/node/mini_miner.cpp:
```diff
@@ -71,8 +71,11 @@ MiniMiner::MiniMiner(const CTxMemPool& mempool, const std::vector<COutPoint>& ou
 
     // Add every entry to m_entries_by_txid and m_entries, except the ones that will be replaced.
```
Changes in this commit is identical to the one in https://github.com/bitcoin/bitcoin/pull/28762/commits/d525998b07be066c97a36536b58f0b98d816be39 , but here the `txmempool.h` dependency from `mini_miner.h`, and the imports are not updated?

sedited committed:
[refactor] Add helper for retrieving mempool entry

In places where the iterator is only needed for accessing the actual
entry, it should not be required to first retrieve the iterator.

TheCharlatan commented:
Updated 105a0f4db4ffdc25d3ad30300c949d46d5d8e647 -> 65839d6267af1f5e0e04aded72cbfa23b56a1237 ([simplifyMemPoolInteractions_4](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_4) -> [simplifyMemPoolInteractions_5](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_5),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_4..simplifyMemPoolInteractions_5))

* Addressed @ismaelsadeeq's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1380032491), pulling in the improved commit from #28762 
* Added a new first commit to introduce a `GetEntry` helper for retrieving mempool entries. This allows avoiding the usage of `GetIter` just for retrieving a mempool entry.
* Use this new `GetEntry` helper over `GetIter` where applicable.
* Refactored the `blockencodings_tests` commit to use a simpler helper function.
* Dropped commit 5633e6770538c83cc5a75cf926347a709ae7e2c0, since `GetEntry` is now used instead.

TheCharlatan commented:
Rebased 65839d6267af1f5e0e04aded72cbfa23b56a1237 -> fc5e4dea6ffa21191e852a79103a2dfc9c241df1 ([simplifyMemPoolInteractions_5](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_5) -> [simplifyMemPoolInteractions_6](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_6),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_5..simplifyMemPoolInteractions_6))

* Dropped the commit that was integrated in #28762 e3b2e630b219ca15fe0b2640ca422712c86ac33d

glozow commented on src/txmempool.h:
```diff
@@ -684,6 +684,8 @@ class CTxMemPool
         return (mapTx.count(gtxid.GetHash()) != 0);
     }
 
+    std::optional<CTxMemPoolEntry::CTxMemPoolEntryRef> GetEntry(const uint256& hash) const LIFETIMEBOUND;
```
Concept ACK to handing out references to the entry instead of iterators into mapTx

ISTM that this should require holding `CTxMemPool::cs`?

TheCharlatan replied:
Yeah, missed that. Will add.

TheCharlatan commented:
Updated fc5e4dea6ffa21191e852a79103a2dfc9c241df1 -> a20c1d2b81a702f97267d4f098f13ed80e9320cf ([simplifyMemPoolInteractions_6](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_6) -> [simplifyMemPoolInteractions_7](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_7),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_6..simplifyMemPoolInteractions_7))

* Addressed @glozow's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1381974204), making the locked mutex a requirement for calling `GetEntry`.

Sjors commented:
The first commit ee79f05f1fef57bfbea24333a46e525385d02319 introduces a dangling reference warning for me. gcc 13.1.0 on Ubuntu 23.04.

```
./configure --enable-suppress-external-warnings --disable-fuzz-binary --without-gui
make
...

validation.cpp: In member function ‘PackageMempoolAcceptResult {anonymous}::MemPoolAccept::AcceptPackage(const Package&, ATMPArgs&)’:
validation.cpp:1488:36: warning: possibly dangling reference to a temporary [-Wdangling-reference]
 1488 |             const CTxMemPoolEntry& entry = Assert(m_pool.GetEntry(txid)).value();
      |                                    ^~~~~
validation.cpp:1488:80: note: the temporary was destroyed at the end of the full expression ‘(&(& inline_assertion_check<true, std::optional<std::reference_wrapper<const CTxMemPoolEntry> > >((&(({anonymous}::MemPoolAccept*)this)->{anonymous}::MemPoolAccept::m_pool)->CTxMemPool::GetEntry((* &(& txid)->transaction_identifier<false>::operator const uint256&())), ((const char*)"validation.cpp"), 1488, ((const char*)(& __func__)), ((const char*)"m_pool.GetEntry(txid)")))->std::optional<std::reference_wrapper<const CTxMemPoolEntry> >::value())->std::reference_wrapper<const CTxMemPoolEntry>::operator const CTxMemPoolEntry&()’
 1488 |             const CTxMemPoolEntry& entry = Assert(m_pool.GetEntry(txid)).value();
      |                                                                                ^
```

Sjors commented on src/txmempool.cpp:
```diff
@@ -850,6 +850,14 @@ std::vector<TxMempoolInfo> CTxMemPool::infoAll() const
+std::optional<CTxMemPoolEntry::CTxMemPoolEntryRef> CTxMemPool::GetEntry(const uint256& hash) const
+{
+    indexed_transaction_set::const_iterator i = mapTx.find(hash);
+    if (i == mapTx.end())
```
ee79f05f1fef57bfbea24333a46e525385d02319: `{`

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -702,9 +702,9 @@ class ChainImpl : public Chain
         LOCK(m_node.mempool->cs);
-        return m_node.mempool->CalculateMemPoolAncestors(entry, limits).has_value();
+        std::string err_string;
+        return m_node.mempool->CheckPackageLimits({tx}, entry.GetTxSize(), err_string);
```
6b6793e6835dd2ddf367490362042bf34eb260a3: The only caller `checkChainLimits` of is the wallet's `CreateTransactionInternal`, which will say "Transaction has too long of a mempool chain" regardless of which limit is hit.

A (wallet) followup should propagate `err_string` up.

Sjors reviewed (commented):
Concept ACK. Lightly reviewed the refactoring (as of a20c1d2b81a702f97267d4f098f13ed80e9320cf) which all looks reasonable, but I'm not very familiar with mempool code.

Sjors commented on src/validation.cpp:
```diff
@@ -1485,9 +1485,8 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
-            auto iter = m_pool.GetIter(txid);
-            assert(iter != std::nullopt);
-            results_final.emplace(wtxid, MempoolAcceptResult::MempoolTx(iter.value()->GetTxSize(), iter.value()->GetFee()));
+            const CTxMemPoolEntry& entry = Assert(m_pool.GetEntry(txid)).value();
```
ee79f05f1fef57bfbea24333a46e525385d02319: this fixes the dangling warning:

```cpp
const CTxMemPoolEntry& entry{Assert(m_pool.GetEntry(txid)).value()};
```

TheCharlatan replied:
Thank, might be a good idea to bracket initialize in other places too.

TheCharlatan commented:
Updated a20c1d2b81a702f97267d4f098f13ed80e9320cf -> 836a0a7566e166556f80303a9b3dd9560587159c ([simplifyMemPoolInteractions_7](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_7) -> [simplifyMemPoolInteractions_8](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_8),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_7..simplifyMemPoolInteractions_8))

* Addressed @Sjors's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1382967058), adding brackets to if statement
* Applied @Sjors's [patch](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1383019317) and also bracket initialize some other vars.

TheCharlatan commented:
Updated 836a0a7566e166556f80303a9b3dd9560587159c -> 11b8735852bd7439c4a07c88d07d7dfb13a3c7b7 ([simplifyMemPoolInteractions_8](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_8) -> [simplifyMemPoolInteractions_9](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_9),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_8..simplifyMemPoolInteractions_9))

* Fixed msvc build error by using `auto` instead of `CTxMemPoolEntry`, which led to implicit conversions.

TheCharlatan commented on src/txmempool.h:
```diff
@@ -684,6 +684,8 @@ class CTxMemPool
         return (mapTx.count(gtxid.GetHash()) != 0);
     }
 
+    std::optional<CTxMemPoolEntry::CTxMemPoolEntryRef> GetEntry(const uint256& hash) const LIFETIMEBOUND EXCLUSIVE_LOCKS_REQUIRED(cs);
```
Yeah, I thought this would better convey its semantics, but with all the force pushes over the past two hours, it seems like it is not intuitive to work with, so I'll revert to using a pointer.

TheCharlatan commented:
Updated 66a4c881e3387b55b26cd2a890bd0d5d522c6727 -> 995aa6b9cb5d1ea685a5d2ac6767d21b373e4c84 ([simplifyMemPoolInteractions_9](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_9) -> [simplifyMemPoolInteractions_10](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_10),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_9..simplifyMemPoolInteractions_10))

* Addressed @maflcko's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1383242173), changed to`CTxMemPoolEntry* GetEntry(...)`.

ismaelsadeeq commented on src/test/miniminer_tests.cpp:
```diff
@@ -179,9 +179,9 @@ BOOST_FIXTURE_TEST_CASE(miniminer_1p1c, TestChain100Setup)
     };
```
nit
https://github.com/bitcoin/bitcoin/blob/6903d8aa37e01d16aea0b53c3729f8f92c7d2bf0/src/test/miniminer_tests.cpp#L97
Can also be
```cpp        
BOOST_CHECK(negative_modified_fees < feerate_zero.GetFee(pool.GetEntry(tx_mod_negative->GetHash())->GetTxSize()));
```

ismaelsadeeq reviewed (approved):
re ACK 995aa6b9cb5d1ea685a5d2ac6767d21b373e4c84

glozow commented on src/txmempool.h:
```diff
@@ -684,6 +684,8 @@ class CTxMemPool
         return (mapTx.count(gtxid.GetHash()) != 0);
     }
 
+    const CTxMemPoolEntry* GetEntry(const uint256& hash) const LIFETIMEBOUND EXCLUSIVE_LOCKS_REQUIRED(cs);
```
6903d8aa37e01d16aea0b53c3729f8f92c7d2bf0
Could you use `Txid` instead? or `GenTxid`

glozow commented on src/test/blockencodings_tests.cpp:
```diff
@@ -62,7 +62,7 @@ BOOST_AUTO_TEST_CASE(SimpleRoundTripTest)
     LOCK2(cs_main, pool.cs);
     pool.addUnchecked(entry.FromTx(block.vtx[2]));
-    BOOST_CHECK_EQUAL(pool.mapTx.find(block.vtx[2]->GetHash())->GetSharedTx().use_count(), SHARED_TX_OFFSET + 0);
+    BOOST_CHECK_EQUAL(pool.get(block.vtx[2]->GetHash()).use_count(), SHARED_TX_OFFSET + 0);
```
Wow TIL there's a `get`!

glozow commented on src/node/miner.h:
```diff
@@ -142,7 +142,7 @@ class BlockAssembler
     uint64_t nBlockSigOpsCost;
     CAmount nFees;
-    CTxMemPool::setEntries inBlock;
+    std::unordered_set<uint256, SaltedTxidHasher> inBlock;
```
aedfef28af6e8c5bd8833fb63e2ef4f3264f934c
could use `Txid` for this as well

glozow commented on src/node/miner.cpp:
```diff
@@ -298,7 +298,7 @@ void BlockAssembler::addPackageTxs(const CTxMemPool& mempool, int& nPackagesSele
     indexed_modified_transaction_set mapModifiedTx;
     // Keep track of entries that failed inclusion, to avoid duplicate work
-    CTxMemPool::setEntries failedTx;
+    std::set<uint256> failedTx;
```
same

glozow reviewed (commented):
LGTM 995aa6b9cb5d1ea685a5d2ac6767d21b373e4c84, just a couple introductions of uint256 that should probably be the type-safe ones instead

glozow reviewed (commented):
drahty pong

TheCharlatan commented:
Updated 995aa6b9cb5d1ea685a5d2ac6767d21b373e4c84 -> 3e7595b11bdad260efb39adc42677ed0beae186d ([simplifyMemPoolInteractions_10](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_10) -> [simplifyMemPoolInteractions_11](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_11),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_10..simplifyMemPoolInteractions_11))

* Addressed @ismaelsadeeq's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1386552287), adding lock assertion.
* Addressed @ismaelsadeeq's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1386609897), applying suggestion replacing `GetIter` with `GetEntry`.
* Addressed @glozow's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1386805310), making `GetEntry` accept a `Txid` type.
* Addressed @glozow's [comment_1](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1386820528) and [comment_2](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1386821034), applying the `Txid` type to the `inBlock` and `failedTx` containers.

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -344,10 +343,12 @@ UniValue MempoolToJSON(const CTxMemPool& pool, bool verbose, bool include_mempoo
-        for (const CTxMemPoolEntry& e : pool.mapTx) {
-            const uint256& hash = e.GetTx().GetHash();
+        for (const auto& txinfo : pool.infoAll()) {
+            const uint256& hash = txinfo.tx->GetHash();
+            const auto entry{pool.GetEntry(Txid::FromUint256(hash))};
```
nit 65c4b3bac0894dfdf7e1c756697f9dd0da1f97e7: don't need to go to uint256 and back again
```suggestion
            const Txid& hash = txinfo.tx->GetHash();
            const auto entry{pool.GetEntry(hash)};
```

glozow reviewed (commented):
ACK 3e7595b11bdad260efb39adc42677ed0beae186d

TheCharlatan commented:
Updated 3e7595b11bdad260efb39adc42677ed0beae186d -> 2be5e799ba623f969f5ffc59667a5bc6799df290 ([simplifyMemPoolInteractions_11](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_11) -> [simplifyMemPoolInteractions_12](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_12),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_11..simplifyMemPoolInteractions_12))

* Addressed @glozow's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1387067811), dropping redundant conversion.

glozow commented:
reACK 2be5e799ba623f969f5ffc59667a5bc6799df290

ismaelsadeeq reviewed (approved):
ACK 2be5e799ba623f969f5ffc59667a5bc6799df290 🥘

stickies-v commented on src/blockencodings.cpp:
```diff
@@ -107,12 +107,13 @@ ReadStatus PartiallyDownloadedBlock::InitData(const CBlockHeaderAndShortTxIDs& c
     LOCK(pool->cs);
-    for (size_t i = 0; i < pool->vTxHashes.size(); i++) {
-        uint64_t shortid = cmpctblock.GetShortID(pool->vTxHashes[i].first);
+    for (size_t i = 0; i < pool->txns_randomized.size(); i++) {
+        const auto& tx = pool->txns_randomized[i];
```
nit (although I realize this diff is from a scripted-diff, so may not be worth updating here)
```suggestion
    for (const auto& tx : pool->txns_randomized) {
```

stickies-v reviewed (commented):
Code Review ACK 2be5e799ba623f969f5ffc59667a5bc6799df290

Left a number of suggestions: some about performance improvements, some about avoiding UB, and some general style/concise nits, but they can all be done in a follow-up too.

stickies-v committed:
[refactor] Add helper for iterating through mempool entries

Instead of reaching into the mapTx data structure, use a helper method
that provides the required vector of CTxMemPoolEntry pointers.

TheCharlatan commented on src/node/interfaces.cpp:
```diff
@@ -806,8 +807,8 @@ class ChainImpl : public Chain
         if (!m_node.mempool) return;
         LOCK2(::cs_main, m_node.mempool->cs);
-        for (const CTxMemPoolEntry& entry : m_node.mempool->mapTx) {
-            notifications.transactionAddedToMempool(entry.GetSharedTx());
+        for (const auto& txinfo : m_node.mempool->infoAll()) {
+            notifications.transactionAddedToMempool(txinfo.tx);
```
Helpers are good, so I'll add this patch as a  separate commit.

TheCharlatan commented:
Thank you for the review @stickies-v!

Updated 2be5e799ba623f969f5ffc59667a5bc6799df290 -> fc710d51b8a64d718dd6d1f66c5eb092cb5f86e9 ([simplifyMemPoolInteractions_12](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_12) -> [simplifyMemPoolInteractions_13](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_13),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_12..simplifyMemPoolInteractions_13))

* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1387953875), adding a new first commit implementing the suggested patch for a more efficient helper method for iterating through mempool transactions. The affected lines now retrieve a vector of `CTxMemPoolEntry*` through this new `entryAll()` instead of calling `infoAll()`.
* Addressed @stickies-v's [comment_1](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1387904513), [comment_2](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1387907915), [comment_3](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1387927635), making assertion checks more compact and adding some in the tests for clarity where required.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1387913882), adding a brief rationale to the message of the commit introducing `vTxHashes` as a vector of `CTransactionRef`.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1387920369), making `GetEntry` method body more concise.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1388011828), simplifying looping over `GetMemPoolChildrenConst()`.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1388021604), explicitly calling `value()` on the option returned by `GetIter(...)`.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1388039482), replacing the `get_iter_from_wtxid` call in `removeUnchecked` with the more widely used `GetIter`. 
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1388044605), replacing the for loop iterating over `txns_randomized` in `blockencodings.cpp` with a range based for loop.

TheCharlatan commented on src/txmempool.h:
```diff
@@ -695,6 +697,7 @@ class CTxMemPool
     /** Returns info for a transaction if its entry_sequence < last_sequence */
     TxMempoolInfo info_for_relay(const GenTxid& gtxid, uint64_t last_sequence) const;
 
+    std::vector<const CTxMemPoolEntry*> entryAll() const EXCLUSIVE_LOCKS_REQUIRED(cs);
```
Thanks, will apply.

TheCharlatan commented:
Thank you for the patch @maflcko,

Updated f2d8e17851da215d0b85ffad9d80e50eb7b2b363 -> 91df77fde5adc7e2e785aa0d60b82a8515c9b538 ([simplifyMemPoolInteractions_13](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_13) -> [simplifyMemPoolInteractions_14](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_14),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_13..simplifyMemPoolInteractions_14))

* Applied @maflcko's [patch](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1389223982), making `entryAll()` return references to `CTxMemPoolEntry` instead of pointers.

TheCharlatan commented on src/node/interfaces.cpp:
```diff
@@ -806,8 +805,8 @@ class ChainImpl : public Chain
     {
         if (!m_node.mempool) return;
         LOCK2(::cs_main, m_node.mempool->cs);
```
This seems a bit out of scope of this PR, but I think you are right, I don't see the need for `::cs_main` here either. Maybe open a separate PR for this?

TheCharlatan commented on src/test/validation_block_tests.cpp:
```diff
@@ -284,7 +284,7 @@ BOOST_AUTO_TEST_CASE(mempool_locks_reorg)
         // Check that all txs are in the pool
         {
             LOCK(m_node.mempool->cs);
```
No, but does the other one in the `while` loop make sense?

TheCharlatan commented:
Will address the outstanding comments shortly.

TheCharlatan commented:
Thank you for the review @maflcko,

Updated 91df77fde5adc7e2e785aa0d60b82a8515c9b538 -> 3b73acb3ec678f5099aa0d77178a0b0d50670c2b ([simplifyMemPoolInteractions_14](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_14) -> [simplifyMemPoolInteractions_15](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_15),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_14..simplifyMemPoolInteractions_15))

* Addressed @maflcko's [comment_1](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1389240397), [comment_2](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1389241558), [comment_3](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1389248596), using `Assert` on assignment to const reference.
* Addressed @maflcko's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1389274834), using `get` to access  transactions in `mempool_tests` instead of `GetIter`.
* Addressed @maflcko's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1389287348), applying `Assert(GetEntry(...))`, instead of `GetIter(...).value()` to turn the potentially thrown exception into an assertion failure.
* Addressed @maflcko's [comment](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1389292338), removing the redundant `mempool.cs` lock.
* Fixed tidy failure.

TheCharlatan commented on src/policy/rbf.cpp:
```diff
@@ -35,8 +36,8 @@ RBFTransactionState IsRBFOptIn(const CTransaction& tx, const CTxMemPool& pool)
     // signaled for RBF if any unconfirmed parents have signaled.
-    const CTxMemPoolEntry entry{*pool.mapTx.find(tx.GetHash())};
-    auto ancestors{pool.AssumeCalculateMemPoolAncestors(__func__, entry, CTxMemPool::Limits::NoLimits(),
+    const auto entry{Assert(pool.GetEntry(tx.GetHash()))};
```
Addressed this, but did not squash, since the change is a bit different to the others.

stickies-v reviewed (approved):
ACK 3b73acb3ec678f5099aa0d77178a0b0d50670c2b

TheCharlatan commented:
Updated 3b73acb3ec678f5099aa0d77178a0b0d50670c2b -> 4dd94ca18f6fc843137ffca3e6d3e97e4f19377b ([simplifyMemPoolInteractions_15](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_15) -> [simplifyMemPoolInteractions_16](https://github.com/TheCharlatan/bitcoin/tree/simplifyMemPoolInteractions_16),  [compare](https://github.com/TheCharlatan/bitcoin/compare/simplifyMemPoolInteractions_15..simplifyMemPoolInteractions_16))

* Applied @stickies-v's [patch](https://github.com/bitcoin/bitcoin/pull/28391#discussion_r1389421589)

TheCharlatan commented on src/kernel/mempool_entry.h:
```diff
@@ -172,8 +172,10 @@ class CTxMemPoolEntry
     mutable Epoch::Marker m_epoch_marker; //!< epoch when last touched, useful for graph algorithms
 };
 
+using CTxMemPoolEntryRef = CTxMemPoolEntry::CTxMemPoolEntryRef;
```
Can it be replaced without declaration order issues?

stickies-v replied:
I tried that at first but couldn't figure a more concise way given the `Parents` and `Children` typedefs.

stickies-v reviewed (approved):
re-ACK 4dd94ca18f6fc843137ffca3e6d3e97e4f19377b

glozow commented:
reACK 4dd94ca

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -836,6 +836,18 @@ static TxMempoolInfo GetInfo(CTxMemPool::indexed_transaction_set::const_iterator
+
+    std::vector<CTxMemPoolEntryRef> ret;
+    ret.reserve(mapTx.size());
+    for (const auto& it : GetSortedDepthAndScore()) {
```
Is there a reason that we are invoking `GetSortedDepthAndScore()` here, ie do any of the callers rely on a particular sort order?

(I'm planning to eliminate this function in #28676, so just want to understand if this behavior needs to be preserved for some reason.)

TheCharlatan replied:
This function is supposed to mimic `infoAll`, which was originally used in this patchset to retrieve mempool information instead of directly accessing `mapTx`. Looking at its usage you are right, the way I read this too there were no order promises made prior to this patch. I opened https://github.com/bitcoin/bitcoin/pull/29019 to instead just iterate through `mapTx`.

glozow commented:
Should we backport 453b4813ebc74859864803e9972b58e4be76a4d6?
